src/sentry/seer/similarity/utils.py

```diff
@@ -52,27 +56,42 @@ def get_stacktrace_string(data: dict[str, Any]) -> str:
                     for frame in exception_value["values"]
                     if frame.get("id") == "frame" and frame.get("contributes")
                 ]
-                num_frames = len(contributing_frames)
-                if frame_count + num_frames > MAX_FRAME_COUNT:
-                    remaining_frame_count = MAX_FRAME_COUNT - frame_count
-                    contributing_frames = contributing_frames[-remaining_frame_count:]
-                    num_frames = remaining_frame_count
-                frame_count += num_frames
+                contributing_frames = _discard_excess_frames(
+                    contributing_frames, MAX_FRAME_COUNT, frame_count
+                )
+                frame_count += len(contributing_frames)
 
                 for frame in contributing_frames:
                     frame_dict = {"filename": "", "function": "", "context-line": ""}
                     for frame_values in frame.get("values", []):
                         if frame_values.get("id") in frame_dict:
                             frame_dict[frame_values["id"]] = _get_value_if_exists(frame_values)
 
-                    frame_str += f'  File "{frame_dict["filename"]}", function {frame_dict["function"]}\n    {frame_dict["context-line"]}\n'
+                    if not _is_snipped_context_line(frame_dict["context-line"]):
+                        found_non_snipped_context_line = True
 
+                    frame_strings.append(
+                        f'  File "{frame_dict["filename"]}", function {frame_dict["function"]}\n    {frame_dict["context-line"]}\n'
+                    )
         # Only exceptions have the type and value properties, so we don't need to handle the threads
         # case here
-        if exception.get("id") == "exception":
-            stacktrace_str += f"{exc_type}: {exc_value}\n"
-        if frame_str:
-            stacktrace_str += frame_str
+        header = f"{exc_type}: {exc_value}\n" if exception["id"] == "exception" else ""
+
+        result_parts.append((header, frame_strings))
+
+    # For performance reasons, if the entire stacktrace is made of minified frames, lower the
+    # maximum, since minified frames are significantly more token-dense than non-minified ones
+    result_max_frames = (
+        MAX_FRAME_COUNT
+        if found_non_snipped_context_line
+        else FULLY_MINIFIED_STACKTRACE_MAX_FRAME_COUNT
+    )
+    final_frame_count = 0
+
+    for header, frame_strings in result_parts:
+        frame_strings = _discard_excess_frames(frame_strings, result_max_frames, final_frame_count)
```

Comment:
would this extra `_discard_excess_frames` call only need to occur if `found_non_snipped_context_line = True`? otherwise the call on line 59 should truncate to `MAX_FRAME_COUNT` already?

Reply:
I put the conditional on the max because it seemed simpler (and `_discard_excess_frames` short-circuits pretty quickly), but you're right, I could move it here. Lemme do that and re-run the tests.